File: backend/app/schemas/calendar.py
from datetime import date, datetime, time
from typing import Optional

from pydantic import BaseModel

from app.models.calendar_block import BlockStatus


class BlockNoteResponse(BaseModel):
    id: str
    content: str
    source: str
    created_at: datetime

    model_config = {"from_attributes": True}


class TaggedTaskSummary(BaseModel):
    id: str
    title: str
    status: str
    priority: str

    model_config = {"from_attributes": True}


class CalendarBlockResponse(BaseModel):
    id: str
    task_id: Optional[str] = None
    scheduled_date: date
    start_time: time
    end_time: time
    duration_minutes: int
    status: BlockStatus
    is_locked: bool
    source: str = "alfred"
    title: Optional[str] = None
    # Denormalized task fields for rendering
    task_title: str
    task_priority: str
    task_energy_level: Optional[str]
    task_status: str

    model_config = {"from_attributes": True}


class BlockDetailResponse(CalendarBlockResponse):
    """Extended response with notes and tagged tasks."""
    notes: list[BlockNoteResponse] = []
    tagged_tasks: list[TaggedTaskSummary] = []


class CalendarBlockCreate(BaseModel):
    task_id: str
    scheduled_date: date
    start_time: time
    end_time: time
    is_locked: bool = True


class CalendarBlockUpdate(BaseModel):
    scheduled_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    is_locked: Optional[bool] = None
    status: Optional[BlockStatus] = None
    title: Optional[str] = None


class BlockNoteCreate(BaseModel):
    content: str


class BlockTagRequest(BaseModel):
    task_id: str


class WeekScheduleRequest(BaseModel):
    week_start: date
    force_reschedule: bool = False


class WeekScheduleResponse(BaseModel):
    blocks: list[CalendarBlockResponse]
    week_start: date
    week_end: date
    tasks_scheduled: int
    tasks_unschedulable: int
    unschedulable_task_ids: list[str]

File: frontend/src/types/calendar.ts
import type { Priority, EnergyLevel, TaskStatus } from './task'

export type BlockStatus = 'scheduled' | 'confirmed' | 'completed' | 'cancelled'

export interface BlockNote {
  id: string
  content: string
  source: string
  created_at: string
}

export interface TaggedTask {
  id: string
  title: string
  status: TaskStatus
  priority: Priority
}

export interface CalendarBlock {
  id: string
  task_id: string | null
  scheduled_date: string
  start_time: string
  end_time: string
  duration_minutes: number
  status: BlockStatus
  is_locked: boolean
  source?: string
  title?: string | null
  task_title: string
  task_priority: Priority
  task_energy_level: EnergyLevel | null
  task_status: TaskStatus
  notes?: BlockNote[]
  tagged_tasks?: TaggedTask[]
}

export interface WeekScheduleResponse {
  blocks: CalendarBlock[]
  week_start: string
  week_end: string
  tasks_scheduled: number
  tasks_unschedulable: number
  unschedulable_task_ids: string[]
}

export interface WeekScheduleRequest {
  week_start: string
  force_reschedule?: boolean
}
